api/capi/include/nntrainer.h

```diff
@@ -215,6 +215,51 @@ int ml_train_model_set_optimizer(ml_train_model_h model,
 int ml_train_model_set_dataset(ml_train_model_h model,
                                ml_train_dataset_h dataset);
 
+/**
+ * @brief Save states of the model after training
+ * @details Use this function to save the current states of the model. @a States
+ * refers to the data which is used for training such as weights, trainable
+ * parameters, optimizer variables, etcs. To save the model, the model
+ * handle must be compiled and have actual allocated memory of states to save.
+ *
+ * @since_tizen 6.5
+ * @remarks  This function does not saves the model architecture but only saves
+ * the states.
+ * @remarks @a file_version states the internal format version starting from 0
+ * @param[in] model The NNTrainer model handle to save
+ * @param[in] path  Pth to save the file
+ * @param[in] file_version file format version
+ * @return @c 0 on success, Otherwise a negative error value
+ * @retval #ML_ERROR_NONE Successful.
+ * @retval #ML_ERROR_INVALID_PARAMETER File path is invalid or model does not
+ * have any allocated states.
+ * @retval #ML_ERROR_NOT_SUPPORTED Not supported format.
+ */
+int ml_train_model_save_states(ml_train_model_h model, const char *path,
+                               int file_version);
```

Comment:
There are three options regarding file_verison.

option 1. current PR (using `int file_version`)
option 2. using dedicated enum such as `ML_TRAIN_STATES_FILE_FORMAT`
option 3. we rely on the extension of the `path`

option 3. is simpler and intuitive but there is one big caveat that if we are having extension crash across versions, it might be impossible to deal with.

Reply:
`save_states` goes against the extension of the path.
`.tflite` saves the states as well as the model architecture while `.bin` + `.ini` has different behavior.

Reply:
The same issue exists with `ml_train_model_load`. When using with `tflite` format, does the API validates the model architecture or overwrites the model with the given architecture from `tflite` format.

Reply:
We will need to save states only. This is also what tflite guys are trying to do.

See below pseudo code from their side.
```python
## This code is from https://github.com/tensorflow/community/pull/390
# Construct the interpreter and get the signature runners.
interpreter = tf.lite.Interpreter(TFLITE_FILE_PATH)
train = interpreter.get_signature_runner(method_name="train")
infer = interpreter.get_signature_runner(method_name="infer")
save = interpreter.get_signature_runner(method_name="save")
restore = interpreter.get_signature_runner(method_name="restore")
# Restore trainable weights from files
restore("/path/to/checkpoints")
# Run training for a few steps
train(train_input, train_labels)
# Save trainable weights back to checkpoint files files
save("/path/to/checkpoints")
# Developers can also use the trained model to run inference now
prediction = infer(input)
```

so states here maps to checkpoint, checkpoint will not have any model architecture saved but only weights.

It's a bit off the topic but, checking the tflite file, we can use tflite format directly to save the checkpoint (This might not be a good idea though).

Reply:
> The same issue exists with ml_train_model_load. When using with tflite format, does the API validates the model architecture or overwrites the model with the given architecture from tflite format. 

I think it's better to use different extension than`*.tflite` although it's possible.

Reply:
How about `ml_train_model_save_checkpoint`?

Reply:
> We will need to save states only. This is also what tflite guys are trying to do.

I think you mean sort of update the states (as the output tflite file should be able to run for inference standalone)?

> so states here maps to checkpoint, checkpoint will not have any model architecture saved but only weights.

This confuses me a bit, then inference would require two files as this checkpoint will not have model architecture?

> It's a bit off the topic but, checking the tflite file, we can use tflite format directly to save the checkpoint (This might not be a good idea though).

not so sure about this one, but thats a good point to raise.